### PR TITLE
[Network] Fix Networking command issues and refactor folded arguments

### LINKED ISSUE #920: az network nic create gives confusing/incorrect error message
Repro: 

Create new (empty) resource group: 
  az resource group create -l westus -n reprotestgroup

Add a nic to the group:
  az network nic create -g reprotestgroup -n testnic --subnet somesubnet

Result:
Error ... Referenced resource /subscriptions/.../resourceGroups/reprotestgroup/..../virtualNetworks/_None_/.../subnets/somesubnet does not exist...

First comment; it is not clear that you have to specify the name of an existing virtual network if you specify a name and not an id for subnet. Neither the usage string nor the help for --subnet tells you this. You have to look at the help for the optional --vnet-name parameter to see under what circumstances it is required. At a minimum, this description should be on --subnet parameter since that is the parameter that introduced the constraint.

Second comment: The usage string should make clear the dependency between the parameters.

Third comment: There is no validation for missing vnet names in the case of providing an name for --subnet. Instead it uses "None" as the name. Very unlikely that this is what the user intended.

### LINKED ISSUE #1237: SDN create NIC problem
When using the following command 
  **az network nic create** 

With required fields by documentation 
```
 az network nic create -g "IMDS_Demo" --subnet "Mysubnet" -n "MyNIC"
Referenced resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/virtualNetworks/None/subnets/Mysubnet does not exist. Please create the required resource and try again.
```

With subnet ID instead
```
az network nic create -g "IMDS_Demo" --subnet "/subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/virtualNetworks/MyVnet2/subnets/Mysubnet" -n "MyNIC"
At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details. {
  "error": {
    "code": "InvalidResourceReference",
    "message": "Resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_DEMO/providers/Microsoft.Network/virtualNetworks/MYVNET2 referenced by resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/networkInterfaces/MyNIC was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.",
    "details": [
      {
        "code": "NotFound",
        "message": "Resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_DEMO/providers/Microsoft.Network/virtualNetworks/MYVNET2 not found."
      }
    ]
  }
}  Correlation ID: 18a90c60-1258-4a63-aad5-18057df8c6c2
```

With Vnet name 
```
az network nic create -g "IMDS_Demo" -n "MyNIC" --subnet "Mysubnet" --vnet-name "MyVnet2"
At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details. {
  "error": {
    "code": "InvalidResourceReference",
    "message": "Resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_DEMO/providers/Microsoft.Network/virtualNetworks/MYVNET2 referenced by resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/networkInterfaces/MyNIC was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.",
    "details": [
      {
        "code": "NotFound",
        "message": "Resource /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_DEMO/providers/Microsoft.Network/virtualNetworks/MYVNET2 not found."
      }
    ]
  }
}  Correlation ID: 04dc6b21-1fbc-4bf3-9060-11fcdb0d3879
```

Workaround to make it work location must be provided
```
az network nic create -g "IMDS_Demo" -n "MyNIC3" --subnet "Mysubnet" --vnet-name "MyVnet2" -l "centralus"
{
  "newNIC": {
    "dnsSettings": {
      "appliedDnsServers": [],
      "dnsServers": []
    },
    "enableIPForwarding": false,
    "ipConfigurations": [
      {
        "etag": "W/\"a17c57d5-3494-4373-80e0-6a9bc85c1566\"",
        "id": "/subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/networkInterfaces/MyNIC3/ipConfigurations/ipconfig1",
        "name": "ipconfig1",
        "properties": {
          "primary": true,
          "privateIPAddress": "10.1.0.5",
          "privateIPAllocationMethod": "Dynamic",
          "provisioningState": "Succeeded",
          "subnet": {
            "id": "/subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/virtualNetworks/MyVnet2/subnets/Mysubnet",
            "resourceGroup": "IMDS_Demo"
          }
        },
        "resourceGroup": "IMDS_Demo"
      }
    ],
    "provisioningState": "Succeeded",
    "resourceGuid": "84408cbd-8a09-4a33-9217-4613e29f2c7a"
  }
}
```

All previous errors are hard to determinate how to fix/work around them, on the documentation -h please add all required and work on the error descriptions

### LINKED ISSUE #1238: SDN create IP Config for NIC problem
When using the following command 
  **az network nic ip-config create** 

With required fields by documentation
az network nic ip-config create -g "IMDS_Demo" --nic-name "MyNICPortal" -n "MyIpConfig"
Subnet reference is required for ipconfiguration /subscriptions/23a27f7b-7bd9-4ef6-91f5-9bf25ea00d6a/resourceGroups/IMDS_Demo/providers/Microsoft.Network/networkInterfaces/MyNICPortal/ipConfigurations/MyIpConfig.

Unable to work around this, tested on CLI and worked fine.

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes #1237, fixes #1238, fixes #920

Replaces instances of `register_folded_cli_argument` in the network module with simpler solutions. Overall, improves the help and validation of the affected parameters while also simplifying the heavily parameterized logic of the old `register_folded_cli_argument`. Once the VM module is refactored, `register_folder_cli_argument` will be removed. 

Also:
- Updates poor quality help text for several networking commands
- Adds --make-primary to network nic ip-config create/update to enable multi-IP config scenarios on NICs.
- Fixes issue with VPN connection deployment template.
- Fixes issue where network nic ip-config create returns the NIC object instead of the IP config object. **Breaking Change**
- Re-enables a VMSS test that had been commented out.
- Updates the `dump_command_table.py` script to work with the recent perf improvements.

### CODE REVIEW

File: src/azure-cli-core/azure/cli/core/commands/template_create.py [hunk 1]
```diff
@@ -110,3 +122,132 @@ def handle_folding(namespace):
...
+    quotes = '""' if platform.system() == 'Windows' else "''"
+
+    if default_none and not allow_none:
+        raise CLIError('Cannot use default_none=True and allow_none=False')
+
+    if not allow_new and not allow_none and not default_none:
+        help_text = 'Name or ID of an existing {}.'.format(display_name)
+    elif not allow_new and allow_none and not default_none:
+        help_text = 'Name or ID of an existing {}, or {} for none.'.format(display_name, quotes)
+    elif allow_new and not allow_none and not default_none:
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/commands/template_create.py]: ` and not default_none` is not needed, as the first if block has excluded this case already.

File: src/azure-cli-core/azure/cli/core/commands/template_create.py [hunk 2]
```diff
@@ -110,3 +122,132 @@ def handle_folding(namespace):
...
+                if value_was_id:
+                    logger = _logging.get_az_logger(__name__)
+                    logger.warning('Ignoring: %s %s', parent_option, parent_val)
+                setattr(namespace, parent_name, None)
+            return # SUCCESS
+
+        # if a parent name was required but not specified, raise a usage error
+        if has_parent and not value_was_id and not parent_val and not allow_new:
+            raise ValueError('incorrect usage: {} ID | {} NAME {} NAME'.format(
+                property_option, property_option, parent_option))
```
Comment (yugangw-msft) [inline src/azure-cli-core/azure/cli/core/commands/template_create.py]: you can use `'incorrect usage: {0} ID | {0} NAME ...` to save an occurrence of `property_option`

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py [hunk 3]
```diff
@@ -289,37 +299,76 @@ def process_ag_url_path_map_rule_create_namespace(namespace): # pylint: disable=
...
 
 def process_lb_create_namespace(namespace):
-    if namespace.public_ip_dns_name:
-        namespace.dns_name_type = 'new'
 
     if namespace.subnet and namespace.public_ip_address:
-        raise argparse.ArgumentError(
-            None, 'Must specify a subnet OR a public IP address, not both.')
+        raise ValueError(
+            'incorrect usage: --subnet NAME_OR_ID [--vnet-name NAME] | --public-ip NAME_OR_ID')
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: Should it be `'incorrect usage: --subnet NAME --vnet-name NAME | --subnet ID | --public-ip NAME_OR_ID`?

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py [hunk 4]
```diff
@@ -289,37 +299,76 @@ def process_ag_url_path_map_rule_create_namespace(namespace): # pylint: disable=
...
+        namespace.subnet = None
+        namespace.virtual_network_name = None
 
-    if not namespace.network_security_group:
-        namespace.network_security_group_type = 'none'
+def process_lb_frontend_ip_namespace(namespace):
+
+    if namespace.subnet and namespace.public_ip_address:
+        raise ValueError(
+            'incorrect usage: --subnet NAME_OR_ID [--vnet-name NAME] | --public-ip NAME_OR_ID')
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: same here

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/recordings/test_vm_scaleset_nics.yaml [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
...4OSwiYWNyIjoiMSIsImFtciI6WyJwd2QiXSwiYXBwaWQiOiIwNGIwNzc5NS04ZGRiLTQ2MWEtYmJlZS0wMmY5ZTFiZjdiNDYiLCJhcHBpZGFjciI6IjAiLCJlX2V4cCI6MTA4MDAsImZhbWlseV9uYW1lIjoiQWRtaW4yIiwiZ2l2ZW5fbmFtZSI6IkFkbWluMiIsImdyb3VwcyI6WyJlNGJiMGI1Ni0xMDE0LTQwZjgtODhhYi0zZDhhOGNiMGUwODYiLCI2Yjk3NzYxYS1kN2QwLTQ4ZjYtYWQ1Ni1mMzhkMzI3Yzg1NTMiXSwiaXBhZGRyIjoiMTY3LjIyMC4wLjE4NiIsIm5hbWUiOiJBZG1pbjIiLCJvaWQiOiI1OTYzZjUwYy03YzQzLTQwNWMtYWY3ZS01MzI5NGRlNzZhYmQiLCJwbGF0ZiI6IjE0IiwicHVpZCI6IjEwMDNCRkZEOTU5Rjg0MjMiLCJzY3AiOiJ1c2VyX2ltcGVyc29uYXRpb24iLCJzdWIiOiJzRGdleFJ3Q05JZlktaHpRampDRHZaVDdJemRmbzRTeXJyNHgwZEROelI0IiwidGlkIjoiNTQ4MjZiMjItMzhkNi00ZmIyLWJhZDktYjdiOTNhM2U5YzVhIiwidW5pcXVlX25hbWUiOiJhZG1pbjJAQXp1cmVTREtUZWFtLm9ubWljcm9zb2Z0LmNvbSIsInVwbiI6ImFkbWluMkBBenVyZVNES1RlYW0ub25taWNyb3NvZnQuY29tIiwidmVyIjoiMS4wIiwid2lkcyI6WyI2MmU5MDM5NC02OWY1LTQyMzctOTE5MC0wMTIxNzcxNDVlMTAiXX0.ByWZCGTYEoUyY8abc1y-td4_jraXQzd9hTEj6QPB31ZB87OHRaCFM204G6xSgGJfi-TrjS7nTS9JYBGIQUnJwjghZEPLgacjO-YxAo2gYMMLEHsT-r3wnh3MMxsJvNZg0VffXDIbSVl3B23_QiNJ-vUmfFVGeMwuNDArhTMzII4P-o4dzmr3As_pL1qR_6ZBcYrPZO7iQtfc_4zvu_a_1_XUoHIbq9gT-_xYWIMNJ5tTkuEBf26K4CMB7fJLEC5U19AVme8lySwzFUtZJs9pF7icAfUBZIyttjPp1yc15nuck1LBjiWE68LrDfNAw1DfACkKgKV2kjseVXWCiPV3TQ]
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/recordings/test_vm_scaleset_nics.yaml]: scrube this

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py (hunk 3 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/_validators.py]: Yes, I like that even better.

**Pull request merged**